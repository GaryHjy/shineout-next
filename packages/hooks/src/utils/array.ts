export const flattenArray = <T>(arr1: any[]): T[] =>
  arr1.reduce(
    (acc, val) => (Array.isArray(val) ? acc.concat(flattenArray(val)) : acc.concat(val)),
    [],
  );
