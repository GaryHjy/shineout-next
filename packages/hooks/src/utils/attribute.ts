export const getDataAttribute = (attrs: Record<string, string | undefined>) => {
  return Object.keys(attrs).reduce((acc, key) => {
    const ns = `data-soui-${key}`;
    if (attrs[key] !== undefined) {
      acc[ns] = attrs[key]!;
    }
    return acc;
  }, {} as Record<string, string>);
};

export const getDataAttributeName = (name: string) => {
  return `data-soui-${name}`;
};
