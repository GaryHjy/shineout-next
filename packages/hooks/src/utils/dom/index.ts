export { default, default as NormalizeWheel } from './normalize-wheel';
export { parsePxToNumber, addResizeObserver } from './element';
export type { OldWheelEvent } from './normalize-wheel.type';
