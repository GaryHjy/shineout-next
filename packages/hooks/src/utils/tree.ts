export const getFlattenTree = <T>(
  data: T[],
  childrenKey = 'children' as keyof T,
  wide?: boolean,
) => {
  const arr: T[][] = [];
  const flatten = (list: T[], path: T[]) => {
    list.forEach((item) => {
      const children = item[childrenKey] as any;
      if (children && children.length > 0) {
        const clonedPath = [...path];
        clonedPath.push(item);
        if (wide) arr.push(clonedPath);
        flatten(children, clonedPath);
      } else {
        arr.push([...path, item]);
      }
    });
  };
  flatten(data, []);
  return arr;
};
