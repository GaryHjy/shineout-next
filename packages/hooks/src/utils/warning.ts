const deprecated = (
  prop: string,
  newProp: string,
  component: string,
  extraMessage?: string
) => {
  if (process.env.NODE_ENV !== 'production') {
    console.warn(
      `[shineout] '${prop}' in ${component} component is deprecated, please use '${newProp}' instead. ${extraMessage}`
    );
  }
};

const breakingChange = (message: string) => {
  if (process.env.NODE_ENV !== 'production') {
    console.warn(`[shineout] ${message}`);
  }
};

const devWarn = (message: string) => {
  if (process.env.NODE_ENV !== 'production') {
    console.warn(`[shineout] ${message}`);
  }
};

export const devUseWarning = {
  deprecated,
  warning: devWarn,
  breaking: breakingChange,
};
