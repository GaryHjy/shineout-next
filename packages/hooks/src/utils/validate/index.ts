import { ObjectType } from '../../common/type';
import { FormItemRule } from '../rule/rule.type';
import getRule from './get-rule';
import { flattenArray } from '../array';
import { wrapFormError } from './error';
export * from './error';

const handleProps = (props: ObjectType) => {
  return Object.keys(props).reduce((acc, key) => {
    const value = props[key];
    const type = typeof value;
    if (type === 'string' || type === 'number') {
      acc[key] = value;
    }
    return acc;
  }, {} as ObjectType);
};

export const validate = <T>(
  value: T | undefined,
  formData: ObjectType,
  rules: FormItemRule<T>,
  p: ObjectType,
) =>
  new Promise<true | Error>((resolve, reject) => {
    if (!rules.length) {
      resolve(true);
      return;
    }
    const props = handleProps(p);
    const [rule, ...restRule] = flattenArray(rules) as FormItemRule<T>;
    if (!rule) {
      validate(value, formData, restRule, props).then(resolve, reject);
      return;
    }
    const callback = (result: true | Error) => {
      if (result !== true) {
        reject(wrapFormError(result, props));
        return;
      }
      validate(value, formData, restRule, props).then(resolve, reject);
    };

    const fn = getRule(rule, props);
    const cb = fn(value, formData, callback, props);
    if (cb) {
      cb?.then(callback.bind(null, true)).catch((e: Error) => {
        reject(wrapFormError(e, props));
      });
    }
  });

export default validate;
