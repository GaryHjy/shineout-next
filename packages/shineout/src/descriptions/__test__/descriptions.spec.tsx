import { render, cleanup, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import Descriptions from '..';
import { baseTest, createClassName, displayTest, snapshotTest } from '../../tests/utils';
import mountTest from '../../tests/mountTest';
import DescriptionsBase from '../__example__/01-base';
import DescriptionsColums from '../__example__/02-colums';
import DescriptionsAlignment from '../__example__/03-alignment';
import DescriptionsBorder from '../__example__/04-border';
import DescriptionsSize from '../__example__/05-size';
import DescriptionsResponsive from '../__example__/06-responsive';

const SO_PREFIX = 'descriptions';
const originClasses = ['wrapper'];
const originItemClasses = [''];
const { wrapper } = createClassName(SO_PREFIX, originClasses, originItemClasses);

// Mock window.matchMedia for node.js environment
beforeAll(() => {
  window.matchMedia =
    window.matchMedia ||
    function () {
      return {
        matches: false,
        addListener: function () {},
        removeListener: function () {},
      };
    };
});

afterEach(cleanup);

mountTest(<Descriptions />);

describe('Alert[Base]', () => {
  displayTest(Descriptions, 'ShineoutDescriptions');
  baseTest(Descriptions, wrapper);
  snapshotTest(<DescriptionsBase />);
  snapshotTest(<DescriptionsColums />, 'about colums');
  snapshotTest(<DescriptionsAlignment />, 'about alignment');
  snapshotTest(<DescriptionsBorder />, 'about border');
  snapshotTest(<DescriptionsSize />, 'about size');
  snapshotTest(<DescriptionsResponsive />, 'about responsive');
  test('should render default', () => {
    render(<Descriptions />);
    screen.debug();
  });
});
