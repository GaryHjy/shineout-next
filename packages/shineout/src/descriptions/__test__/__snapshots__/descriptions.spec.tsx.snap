// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert[Base] should render correctly  1`] = `
<div>
  <div
    class="so-descriptions-wrapper"
  >
    <div
      class="so-descriptions-header"
    >
      <div
        class="so-descriptions-title"
      >
        User Info
      </div>
    </div>
    <div
      class="so-descriptions-body so-descriptions-inline-horizontal"
    >
      <table
        cellpadding="0"
        cellspacing="0"
        class="so-descriptions-table"
      >
        <tbody>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-inline-table"
            >
              <div
                class="so-descriptions-item"
              >
                <div
                  class="so-descriptions-label-inline"
                >
                  Name
                   :
                </div>
                <div
                  class="so-descriptions-value-inline"
                >
                  Mai Mai
                </div>
              </div>
            </td>
            <td
              class="so-descriptions-inline-table"
            >
              <div
                class="so-descriptions-item"
              >
                <div
                  class="so-descriptions-label-inline"
                >
                  Residence
                   :
                </div>
                <div
                  class="so-descriptions-value-inline"
                >
                  Beijing
                </div>
              </div>
            </td>
            <td
              class="so-descriptions-inline-table"
            >
              <div
                class="so-descriptions-item"
              >
                <div
                  class="so-descriptions-label-inline"
                >
                  Address
                   :
                </div>
                <div
                  class="so-descriptions-value-inline"
                >
                  Yingdu Building,Zhichun Road,Beijing
                </div>
              </div>
            </td>
          </tr>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-inline-table"
            >
              <div
                class="so-descriptions-item"
              >
                <div
                  class="so-descriptions-label-inline"
                >
                  Mobile
                   :
                </div>
                <div
                  class="so-descriptions-value-inline"
                >
                  187-2323-9834
                </div>
              </div>
            </td>
            <td
              class="so-descriptions-inline-table"
              colspan="2"
            >
              <div
                class="so-descriptions-item"
              >
                <div
                  class="so-descriptions-label-inline"
                >
                  Hometown
                   :
                </div>
                <div
                  class="so-descriptions-value-inline"
                >
                  Beijing
                </div>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`Alert[Base] should render correctly about alignment 1`] = `
<div>
  <div
    class="so-descriptions-wrapper"
    style="margin-bottom: 24px;"
  >
    <div
      class="so-descriptions-header"
    >
      <div
        class="so-descriptions-title"
      >
        User Info
      </div>
    </div>
    <div
      class="so-descriptions-body so-descriptions-horizontal"
    >
      <table
        cellpadding="0"
        cellspacing="0"
        class="so-descriptions-table"
      >
        <tbody>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
            >
              Name
            </td>
            <td
              class="so-descriptions-value"
            >
              Mai Mai
            </td>
            <td
              class="so-descriptions-label"
            >
              Residence
            </td>
            <td
              class="so-descriptions-value"
            >
              Beijing
            </td>
            <td
              class="so-descriptions-label"
            >
              Address
            </td>
            <td
              class="so-descriptions-value"
            >
              Yingdu Building,Zhichun Road,Beijing
            </td>
          </tr>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
            >
              Hometown
            </td>
            <td
              class="so-descriptions-value"
            >
              Beijing
            </td>
            <td
              class="so-descriptions-label"
            >
              Mobile
            </td>
            <td
              class="so-descriptions-value"
              colspan="3"
            >
              187-2323-9834
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div
    class="so-descriptions-wrapper"
    style="margin-bottom: 24px;"
  >
    <div
      class="so-descriptions-header"
    >
      <div
        class="so-descriptions-title"
      >
        User Info
      </div>
    </div>
    <div
      class="so-descriptions-body so-descriptions-horizontal"
    >
      <table
        cellpadding="0"
        cellspacing="0"
        class="so-descriptions-table"
      >
        <tbody>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Name
            </td>
            <td
              class="so-descriptions-value"
            >
              Mai Mai
            </td>
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Residence
            </td>
            <td
              class="so-descriptions-value"
            >
              Beijing
            </td>
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Address
            </td>
            <td
              class="so-descriptions-value"
            >
              Yingdu Building,Zhichun Road,Beijing
            </td>
          </tr>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Hometown
            </td>
            <td
              class="so-descriptions-value"
            >
              Beijing
            </td>
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Mobile
            </td>
            <td
              class="so-descriptions-value"
              colspan="3"
            >
              187-2323-9834
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div
    class="so-descriptions-wrapper"
  >
    <div
      class="so-descriptions-header"
    >
      <div
        class="so-descriptions-title"
      >
        User Info
      </div>
    </div>
    <div
      class="so-descriptions-body"
    >
      <table
        cellpadding="0"
        cellspacing="0"
        class="so-descriptions-table"
      >
        <tbody>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
            >
              Name
            </td>
            <td
              class="so-descriptions-label"
            >
              Residence
            </td>
            <td
              class="so-descriptions-label"
            >
              Address
            </td>
          </tr>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-value"
            >
              Mai Mai
            </td>
            <td
              class="so-descriptions-value"
            >
              Beijing
            </td>
            <td
              class="so-descriptions-value"
            >
              Yingdu Building,Zhichun Road,Beijing
            </td>
          </tr>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
            >
              Hometown
            </td>
            <td
              class="so-descriptions-label"
              colspan="2"
            >
              Mobile
            </td>
          </tr>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-value"
            >
              Beijing
            </td>
            <td
              class="so-descriptions-value"
              colspan="2"
            >
              187-2323-9834
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`Alert[Base] should render correctly about border 1`] = `
<div>
  <div
    class="so-descriptions-wrapper"
    style="margin-bottom: 20px;"
  >
    <div
      class="so-descriptions-header"
    >
      <div
        class="so-descriptions-title"
      >
        User Info
      </div>
    </div>
    <div
      class="so-descriptions-body so-descriptions-border so-descriptions-table-layout-fixed so-descriptions-horizontal"
    >
      <table
        cellpadding="0"
        cellspacing="0"
        class="so-descriptions-table"
      >
        <tbody>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
            >
              Name
            </td>
            <td
              class="so-descriptions-value"
            >
              Mai Mai
            </td>
            <td
              class="so-descriptions-label"
            >
              Residence
            </td>
            <td
              class="so-descriptions-value"
            >
              Beijing
            </td>
            <td
              class="so-descriptions-label"
            >
              Address
            </td>
            <td
              class="so-descriptions-value"
            >
              Yingdu Building,Zhichun Road,Beijing
            </td>
          </tr>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
            >
              Mobile
            </td>
            <td
              class="so-descriptions-value"
            >
              187-2323-9834
            </td>
            <td
              class="so-descriptions-label"
            >
              Hometown
            </td>
            <td
              class="so-descriptions-value"
              colspan="3"
            >
              Beijing
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div
    class="so-descriptions-wrapper"
    style="margin-bottom: 20px;"
  >
    <div
      class="so-descriptions-header"
    >
      <div
        class="so-descriptions-title"
      >
        User Info
      </div>
    </div>
    <div
      class="so-descriptions-body so-descriptions-border so-descriptions-table-layout-fixed so-descriptions-horizontal"
    >
      <table
        cellpadding="0"
        cellspacing="0"
        class="so-descriptions-table"
      >
        <tbody>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Name
            </td>
            <td
              class="so-descriptions-value"
            >
              Mai Mai
            </td>
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Residence
            </td>
            <td
              class="so-descriptions-value"
            >
              Beijing
            </td>
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Address
            </td>
            <td
              class="so-descriptions-value"
            >
              Yingdu Building,Zhichun Road,Beijing
            </td>
          </tr>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Mobile
            </td>
            <td
              class="so-descriptions-value"
            >
              187-2323-9834
            </td>
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Hometown
            </td>
            <td
              class="so-descriptions-value"
              colspan="3"
            >
              Beijing
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div
    class="so-descriptions-wrapper"
  >
    <div
      class="so-descriptions-header"
    >
      <div
        class="so-descriptions-title"
      >
        User Info
      </div>
    </div>
    <div
      class="so-descriptions-body so-descriptions-border so-descriptions-table-layout-fixed"
    >
      <table
        cellpadding="0"
        cellspacing="0"
        class="so-descriptions-table"
      >
        <tbody>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
            >
              Name
            </td>
            <td
              class="so-descriptions-label"
            >
              Residence
            </td>
            <td
              class="so-descriptions-label"
            >
              Address
            </td>
          </tr>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-value"
            >
              Mai Mai
            </td>
            <td
              class="so-descriptions-value"
            >
              Beijing
            </td>
            <td
              class="so-descriptions-value"
            >
              Yingdu Building,Zhichun Road,Beijing
            </td>
          </tr>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
            >
              Mobile
            </td>
            <td
              class="so-descriptions-label"
              colspan="2"
            >
              Hometown
            </td>
          </tr>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-value"
            >
              187-2323-9834
            </td>
            <td
              class="so-descriptions-value"
              colspan="2"
            >
              Beijing
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`Alert[Base] should render correctly about colums 1`] = `
<div>
  <div
    class="so-descriptions-wrapper"
    style="margin-bottom: 24px;"
  >
    <div
      class="so-descriptions-header"
    >
      <div
        class="so-descriptions-title"
      >
        User Info
      </div>
    </div>
    <div
      class="so-descriptions-body so-descriptions-horizontal"
    >
      <table
        cellpadding="0"
        cellspacing="0"
        class="so-descriptions-table"
      >
        <tbody>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
            >
              Name
            </td>
            <td
              class="so-descriptions-value"
            >
              Mai Mai
            </td>
          </tr>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
            >
              Residence
            </td>
            <td
              class="so-descriptions-value"
            >
              Beijing
            </td>
          </tr>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
            >
              Address
            </td>
            <td
              class="so-descriptions-value"
            >
              Yingdu Building,Zhichun Road,Beijing
            </td>
          </tr>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
            >
              Mobile
            </td>
            <td
              class="so-descriptions-value"
            >
              187-2323-9834
            </td>
          </tr>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
            >
              Hometown
            </td>
            <td
              class="so-descriptions-value"
            >
              Beijing
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div
    class="so-descriptions-wrapper"
  >
    <div
      class="so-descriptions-header"
    >
      <div
        class="so-descriptions-title"
      >
        User Info
      </div>
    </div>
    <div
      class="so-descriptions-body so-descriptions-horizontal"
    >
      <table
        cellpadding="0"
        cellspacing="0"
        class="so-descriptions-table"
      >
        <tbody>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Name
            </td>
            <td
              class="so-descriptions-value"
            >
              Mai Mai
            </td>
          </tr>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Residence
            </td>
            <td
              class="so-descriptions-value"
            >
              Beijing
            </td>
          </tr>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Address
            </td>
            <td
              class="so-descriptions-value"
            >
              Yingdu Building,Zhichun Road,Beijing
            </td>
          </tr>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Mobile
            </td>
            <td
              class="so-descriptions-value"
            >
              187-2323-9834
            </td>
          </tr>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Hometown
            </td>
            <td
              class="so-descriptions-value"
            >
              Beijing
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`Alert[Base] should render correctly about responsive 1`] = `
<div>
  <div
    class="so-descriptions-wrapper"
  >
    <div
      class="so-descriptions-header"
    >
      <div
        class="so-descriptions-title"
      >
        User Info
      </div>
    </div>
    <div
      class="so-descriptions-body so-descriptions-border so-descriptions-horizontal"
    >
      <table
        cellpadding="0"
        cellspacing="0"
        class="so-descriptions-table"
      >
        <tbody>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Name
            </td>
            <td
              class="so-descriptions-value"
            >
              Mai Mai
            </td>
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Residence
            </td>
            <td
              class="so-descriptions-value"
            >
              Beijing
            </td>
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Address
            </td>
            <td
              class="so-descriptions-value"
            >
              Yingdu Building,Zhichun Road,Beijing
            </td>
          </tr>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Mobile
            </td>
            <td
              class="so-descriptions-value"
            >
              187-2323-9834
            </td>
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Hometown
            </td>
            <td
              class="so-descriptions-value"
              colspan="3"
            >
              Beijing
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`Alert[Base] should render correctly about size 1`] = `
<div>
  <div
    class="so-radio-group so-radio-group-button so-button-group"
    style="margin-bottom: 24px;"
  >
    <button
      class="so-radio-wrapper so-button-secondary so-button-button so-button-outline"
      type="button"
    >
      <input
        style="display: none;"
        tabindex="0"
        type="radio"
      />
      <span>
        small
      </span>
    </button>
    <button
      class="so-radio-wrapper so-radio-wrapper-checked so-button-primary so-button-button so-button-outline"
      type="button"
    >
      <input
        checked=""
        style="display: none;"
        tabindex="0"
        type="radio"
      />
      <span>
        default
      </span>
    </button>
    <button
      class="so-radio-wrapper so-button-secondary so-button-button so-button-outline"
      type="button"
    >
      <input
        style="display: none;"
        tabindex="0"
        type="radio"
      />
      <span>
        large
      </span>
    </button>
  </div>
  <div
    class="so-descriptions-wrapper"
    style="margin-bottom: 20px;"
  >
    <div
      class="so-descriptions-header"
    >
      <div
        class="so-descriptions-title"
      >
        User Info
      </div>
    </div>
    <div
      class="so-descriptions-body so-descriptions-inline-horizontal"
    >
      <table
        cellpadding="0"
        cellspacing="0"
        class="so-descriptions-table"
      >
        <tbody>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-inline-table"
            >
              <div
                class="so-descriptions-item"
              >
                <div
                  class="so-descriptions-label-inline"
                >
                  Name
                   :
                </div>
                <div
                  class="so-descriptions-value-inline"
                >
                  Mai Mai
                </div>
              </div>
            </td>
            <td
              class="so-descriptions-inline-table"
            >
              <div
                class="so-descriptions-item"
              >
                <div
                  class="so-descriptions-label-inline"
                >
                  Residence
                   :
                </div>
                <div
                  class="so-descriptions-value-inline"
                >
                  Beijing
                </div>
              </div>
            </td>
            <td
              class="so-descriptions-inline-table"
            >
              <div
                class="so-descriptions-item"
              >
                <div
                  class="so-descriptions-label-inline"
                >
                  Address
                   :
                </div>
                <div
                  class="so-descriptions-value-inline"
                >
                  Yingdu Building,Zhichun Road,Beijing
                </div>
              </div>
            </td>
          </tr>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-inline-table"
            >
              <div
                class="so-descriptions-item"
              >
                <div
                  class="so-descriptions-label-inline"
                >
                  Mobile
                   :
                </div>
                <div
                  class="so-descriptions-value-inline"
                >
                  187-2323-9834
                </div>
              </div>
            </td>
            <td
              class="so-descriptions-inline-table"
              colspan="2"
            >
              <div
                class="so-descriptions-item"
              >
                <div
                  class="so-descriptions-label-inline"
                >
                  Hometown
                   :
                </div>
                <div
                  class="so-descriptions-value-inline"
                >
                  Beijing
                </div>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div
    class="so-descriptions-wrapper"
  >
    <div
      class="so-descriptions-header"
    >
      <div
        class="so-descriptions-title"
      >
        User Info
      </div>
    </div>
    <div
      class="so-descriptions-body so-descriptions-border so-descriptions-horizontal"
    >
      <table
        cellpadding="0"
        cellspacing="0"
        class="so-descriptions-table"
      >
        <tbody>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Name
            </td>
            <td
              class="so-descriptions-value"
            >
              Mai Mai
            </td>
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Residence
            </td>
            <td
              class="so-descriptions-value"
            >
              Beijing
            </td>
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Address
            </td>
            <td
              class="so-descriptions-value"
            >
              Yingdu Building,Zhichun Road,Beijing
            </td>
          </tr>
          <tr
            class="so-descriptions-row"
          >
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Mobile
            </td>
            <td
              class="so-descriptions-value"
            >
              187-2323-9834
            </td>
            <td
              class="so-descriptions-label"
              style="text-align: right;"
            >
              Hometown
            </td>
            <td
              class="so-descriptions-value"
              colspan="3"
            >
              Beijing
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
